paperqa/agents/models.py

```diff
@@ -0,0 +1,532 @@
+from __future__ import annotations
+
+import asyncio
+import hashlib
+import logging
+import time
+from contextlib import asynccontextmanager
+from datetime import datetime
+from enum import Enum
+from typing import Any, ClassVar
+from uuid import UUID, uuid4
+
+from langchain_core.callbacks import AsyncCallbackHandler
+from langchain_core.messages import BaseMessage, messages_to_dict
+from langchain_core.outputs import ChatGeneration, LLMResult
+from openai import AsyncOpenAI
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    PrivateAttr,
+    ValidationInfo,
+    computed_field,
+    field_validator,
+    model_validator,
+)
+from typing_extensions import Protocol
+
+import paperqa
+from paperqa import (
+    Answer,
+    Context,
+    OpenAILLMModel,
+    PromptCollection,
+    Text,
+    llm_model_factory,
+)
+
+from ..version import __version__
+from .prompts import STATIC_PROMPTS
+
+logger = logging.getLogger(__name__)
+
+
+class SupportsPickle(Protocol):
+    def __reduce__(self) -> str | tuple[Any, ...]: ...
+    def __getstate__(self) -> object: ...
+    def __setstate__(self, state: object) -> None: ...
+
+
+def strip_answer(answer: Answer):
+    """Filter out extra items that do not need to be returned to the user. Modifies in place."""
+    new_contexts = [
+        Context(
+            context=c.context,
+            score=c.score,
+            text=Text(
+                text="",
+                **c.text.model_dump(exclude={"text", "embedding", "doc"}),
+                doc=paperqa.Doc(**c.text.doc.model_dump(exclude={"embedding"})),
+            ),
+        )
+        for c in answer.contexts
+    ]
+    answer.contexts = new_contexts
+
+
+class AgentStatus(str, Enum):
```

Comment:
What is our Python version range? Can we use `enum.StrEnum`?

Reply:
Ahh actually we only require 3.8 -- interesting. I haven't run the test suite with 3.8 but I'm dubious.

Reply:
Gonna come back to this in a later PR — we might need to bump up our minimum python version when we do the merge to main